0000-backtrace-in-core.md

```diff
@@ -0,0 +1,145 @@
+
+Additionally, having this type in `core` will allow its users to provide their own implementations of the backtrace collection and reporting and not rely on the `std`-provided one if they don't want to.
+
+The outcome of this RFC will be a `Backtrace` type in `core` with implementation defined in `std` and compiler-generated implementation when `std` is not linked and user did not provide their own implementation for the reporting functions.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Backtraces are an essential part of the Rust ecosystem, being close coupled with error handling and reporting, be it to the user or the developer. They are usually of a common form:
+```rust
+   Compiling playground v0.0.1 (/playground)
+    Finished dev [unoptimized + debuginfo] target(s) in 1.66s
+     Running `target/debug/playground`
+thread 'main' panicked at 'index out of bounds: the len is 4 but the index is 4', src/main.rs:5:5
+stack backtrace:
+   0: rust_begin_unwind
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/std/src/panicking.rs:493:5
+   1: core::panicking::panic_fmt
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:92:14
+   2: core::panicking::panic_bounds_check
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/panicking.rs:69:5
+   3: <usize as core::slice::index::SliceIndex<[T]>>::index_mut
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/slice/index.rs:190:14
+   4: core::slice::index::<impl core::ops::index::IndexMut<I> for [T]>::index_mut
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/slice/index.rs:26:9
+   5: <alloc::vec::Vec<T,A> as core::ops::index::IndexMut<I>>::index_mut
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/alloc/src/vec/mod.rs:2396:9
+   6: playground::main
+             at ./src/main.rs:5:5
+   7: core::ops::function::FnOnce::call_once
+             at /rustc/53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b/library/core/src/ops/function.rs:227:5
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
+```
+
+They allow for a detailed inspection of what crashed at runtime and which libraries and functions were involved in the process. Obviously reporting and formatting it is also of a big importance, so `Error` allows for backtrace printing with help of a `backtrace()` method.
+
+Currently, `Backtrace` is an essential part of the `std` library and user can control whether the backtrace is enabled or disabled using the environmental variable: `RUST_BACKTRACE`.
+
+In terms of Guide-level changes, there is not much to be discussed - only that it is moved to `core` and if `std` is not linked, automatic backtrace handlers will be generated. Otherwise, the regular implementation of `Backtrace` is present.
+
+TODO: what else should be here?
```

Comment:
I would add a bit more detail about what the Backtrace type is (is it a trait or a struct, etc.), how it is concretely implemented, and how it is used. I would specify where backtrace-related stuff is located (e.g., the collection functions), and give examples of using the Backtrace type (i.e., importing the right things from std and core). I would identify different classes of users (primarily no-std vs std, but also consumers of the type, vs people doing something more complicated such as providing a Backtrace collector). Explain that for most users, the fact that Backtrace is in core doesn't matter since it is re-exported by std. I would write about changes necessary for users who are using Backtrace today.